text/2021-03-11-unified-selector.md

```diff
@@ -2,55 +2,91 @@
+`.Target` selectors in `NetworkChaos`. The controller will iterate this map to
+select every `SelectSpec`. It will construct a unified selector first, and then
+use this selector to `Select` targets. The unified selector may contain a lot of
+implementation inside. The construction would be like:
 
 ```go
 selector := selector.New(selector.SelectorParams{
-	PodSelector: pod.New(r.Client, r.Reader, config.ControllerCfg.ClusterScoped, config.ControllerCfg.TargetNamespace, config.ControllerCfg.AllowedNamespaces, config.ControllerCfg.IgnoredNamespaces),
-	ContainerSelector: container.New(r.Client, r.Reader, config.ControllerCfg.ClusterScoped, config.ControllerCfg.TargetNamespace, config.ControllerCfg.AllowedNamespaces, config.ControllerCfg.IgnoredNamespaces),
+    PodSelector: pod.New(r.Client, r.Reader, config.ControllerCfg.ClusterScoped, config.ControllerCfg.TargetNamespace, config.ControllerCfg.AllowedNamespaces, config.ControllerCfg.IgnoredNamespaces),
+    ContainerSelector: container.New(r.Client, r.Reader, config.ControllerCfg.ClusterScoped, config.ControllerCfg.TargetNamespace, config.ControllerCfg.AllowedNamespaces, config.ControllerCfg.IgnoredNamespaces),
 })
 ```
 
-With the help of dependency injection, it would be constructed easier. The `Selector` method of `PodSelector` would be like:
+With the help of dependency injection, it would be constructed easier. The
+`Selector` method of `PodSelector` would be like:
 
 ```go
 func (impl *PodSelector) Select(ctx context.Context, ps *v1alpha1.PodSelector) ([]interface{}, error)
 ```
 
-The type of second parameter will decide which selector to use. For example, if you want to select with `*v1alpha1.PodSelector`, then the unified selector will use `*PodSelector`, and if you want to select with `*v1alpha1.ContainerSelector`, then the unified selector will use `*ContainerSelector` to select.
+The type of second parameter will decide which selector to use. For example, if
+you want to select with `*v1alpha1.PodSelector`, then the unified selector will
+use `*PodSelector`, and if you want to select with
+`*v1alpha1.ContainerSelector`, then the unified selector will use
+`*ContainerSelector` to select.
 
 The definition of the unified selector would be:
 
 ```go
 func (s *Selector) Select(ctx context.Context, spec interface{}) ([]interface{}, error)
 ```
 
-The controller would construct the `Selector` first, and then use the `Selector` to select the chaos targets. However, as the target may have multiple type (it could be a pod, a container, a volume or an AWS machine), we can only return an `interface{}`, and the implementation of chaos would assert the type by themselves.
+The controller would construct the `Selector` first, and then use the `Selector`
+to select the chaos targets. However, as the target may have multiple type (it
+could be a pod, a container, a volume or an AWS machine), we can only return an
+`interface{}`, and the implementation of chaos would assert the type by
```

Comment:
So the controller does not actually know the type of the target? I wonder how it leads to 

> the controller framework should know the concrete status of every selected targets

Reply:
Oops. It needs to know whether these targets are injected or not (or fault), which can be judged according to the return value of `Apply`/`Recover`. (If `Recover` success, then it must have been in "not injected" status :smile_cat: )

Reply:
So the granularity of success/failure would be at a per-chaos level. What if some of the targets succeeded but the other failed?

Reply:
Nop. In controller, the "success"/"failure" will be  per-"target" level, so that the error handling (like "retry") would be simpler. 

However, to avoid modifying dashboard, I don't want to modify the specification too rudely. So the per-chaos level will be keeped as the "AND" of every target status, and will only be used to show in dashboard.

Reply:
I don't quite get it. How can you get a pre-target status if you do not even know what the target is? (Since it is `interface{}`)

Reply:
The "status" is only about whether this chaos has been executed successfully, which we can get from the following return value of `Apply` or `Recover` function. Why should I know what the target is?

But yes. I have realized that `interface{}` is not enough. It has to provide an ID to record the status/running result.